src/transport.c

```diff
@@ -877,19 +934,58 @@ int _libssh2_transport_send(LIBSSH2_SESSION *session,
         /* Calculate MAC hash. Put the output at index packet_length,
            since that size includes the whole packet. The MAC is
            calculated on the entire unencrypted packet, including all
-           fields except the MAC field itself. */
-        session->local.mac->hash(session, p->outbuf + packet_length,
+           fields except the MAC field itself. This is skipped in the
+           INTEGRATED_MAC case, where the crypto algorithm also does its
+           own hash. */
+        if(!CRYPT_FLAG_R(session, INTEGRATED_MAC))
+          session->local.mac->hash(session, p->outbuf + packet_length,
                                  session->local.seqno, p->outbuf,
                                  packet_length, NULL, 0,
                                  &session->local.mac_abstract);
 
         /* Encrypt the whole packet data, one block size at a time.
-           The MAC field is not encrypted. */
+           The MAC field is not encrypted unless INTEGRATED_MAC. */
+        /* Some crypto back-ends could handle a single crypt() call for
+           encryption, but (presumably) others cannot, so break it up
+           into blocksize-sized chunks to satisfy them all. */
         for(i = 0; i < packet_length; i += session->local.crypt->blocksize) {
             unsigned char *ptr = &p->outbuf[i];
+            size_t bsize = MIN(session->local.crypt->blocksize,
+                                   (int)(packet_length-i));
```

Comment:
```suggestion
            size_t bsize = LIBSSH2_MIN(session->local.crypt->blocksize,
                                   (int)(packet_length-i));
```
Prefer newly added `LIBSSH2_MAX()`/`LIBSSH2_MIN()`.